rollout/trafficrouting/gatewayapi/gatewayapi.go

```diff
@@ -0,0 +1,219 @@
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+)
+
+// Type holds this controller type
+const Type = "GatewayAPI"
+
+const httpRoutes = "httproutes"
+const GatewayAPIUpdateError = "GatewayAPIUpdateError"
+
+var (
+	apiGroupToResource = map[string]string{
+		defaults.DefaultGatewayAPIGroup: httpRoutes,
+	}
+)
+
+type ReconcilerConfig struct {
+	Rollout  *v1alpha1.Rollout
+	Client   ClientInterface
+	Recorder record.EventRecorder
+}
+
+type Reconciler struct {
+	Rollout  *v1alpha1.Rollout
+	Client   ClientInterface
+	Recorder record.EventRecorder
+}
+
+type ClientInterface interface {
+	Get(ctx context.Context, name string, options metav1.GetOptions, subresources ...string) (*unstructured.Unstructured, error)
+	Update(ctx context.Context, obj *unstructured.Unstructured, options metav1.UpdateOptions, subresources ...string) (*unstructured.Unstructured, error)
+}
+
+func (r *Reconciler) sendWarningEvent(id, msg string) {
+	r.sendEvent(corev1.EventTypeWarning, id, msg)
+}
+
+func (r *Reconciler) sendEvent(eventType, id, msg string) {
+	r.Recorder.Eventf(r.Rollout, record.EventOptions{EventType: eventType, EventReason: id}, msg)
+}
+
+func NewDynamicClient(di dynamic.Interface, namespace string) dynamic.ResourceInterface {
+	return di.Resource(GetMappingGVR()).Namespace(namespace)
+}
+
+func GetMappingGVR() schema.GroupVersionResource {
+	return toMappingGVR(defaults.GetGatewayAPIGroupVersion())
+}
+
+func toMappingGVR(apiVersion string) schema.GroupVersionResource {
+	parts := strings.Split(apiVersion, "/")
+	group := parts[0]
+	resourcename := apiGroupToResource[group]
+	version := parts[len(parts)-1]
+	return schema.GroupVersionResource{
+		Group:    group,
+		Version:  version,
+		Resource: resourcename,
+	}
+}
+
+func NewReconciler(cfg *ReconcilerConfig) *Reconciler {
+	reconciler := Reconciler{
+		Rollout:  cfg.Rollout,
+		Client:   cfg.Client,
+		Recorder: cfg.Recorder,
+	}
+	return &reconciler
+}
+
+func (r *Reconciler) UpdateHash(canaryHash, stableHash string, additionalDestinations ...v1alpha1.WeightDestination) error {
+	return nil
+}
+
+func (r *Reconciler) SetWeight(desiredWeight int32, additionalDestinations ...v1alpha1.WeightDestination) error {
+	ctx := context.TODO()
+	rollout := r.Rollout
+	httpRouteName := rollout.Spec.Strategy.Canary.TrafficRouting.GatewayAPI.HTTPRoute
+	httpRoute, err := r.Client.Get(ctx, httpRouteName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	canaryServiceName := rollout.Spec.Strategy.Canary.CanaryService
+	stableServiceName := rollout.Spec.Strategy.Canary.StableService
+	rules, isFound, err := unstructured.NestedSlice(httpRoute.Object, "spec", "rules")
+	if err != nil {
+		return err
+	}
+	if !isFound {
+		return errors.New("spec.rules field was not found in httpRoute")
+	}
+	backendRefs, err := getBackendRefs(rules)
+	if err != nil {
+		return err
+	}
+	if backendRefs == nil {
```

Comment:
Even though this is only checked once  I think along with moving the getService unfound return we could do the same here for backEnd refs not found